url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/KitRobotAgent.java

```diff
@@ -0,0 +1,461 @@
+	private Camera camera;
+	private KitRobotGraphics kitrobotGraphics;
+	private MockGraphics mockgraphics;
+
+	private KitRobotState state;
+
+	private enum KitRobotState {
+		IDLE, HOLDING_KIT, NOT_HOLDING_KIT
+	};
+
+	private final String name;
+
+	/**
+	 * Inner class encapsulates kit and adds states relevant to the stand
+	 * 
+	 * @author dpaje
+	 */
+	public class MyKit {
+		public Kit kit;
+		public KitStatus KS;
+		public int location;
+
+		public MyKit(Kit k) {
+			this.kit = k;
+			this.KS = KitStatus.AWAITING_PICKUP;
+		}
+	}
+
+	public enum KitStatus {
+		AWAITING_PICKUP, REQUESTED, PICKED_UP, ON_STAND, MARKED_FOR_INSPECTION, AWAITING_INSPECTION, FAILED_INSPECTION, PASSED_INSPECTION, SHIPPED;
+	};
+
+	/**
+	 * Constructor for KitRobotAgent class
+	 * 
+	 * @param name
+	 *            name of the kitrobot
+	 */
+	public KitRobotAgent(String name) {
+		super();
+
+		this.name = name;
+		kitWaitingOnConveyor = false;
+		kitRequested = false;
+		numKitsRequested = 0;
+		numKitsToMake = 0;
+		state = KitRobotState.IDLE;
+
+		// Don't assume stand is empty
+		standPositions.put(0, true);
+		standPositions.put(1, false);
+		standPositions.put(2, false);
+	}
+
+	/*
+	 * Messages
+	 */
+
+	@Override
+	public void msgNeedThisManyKits(int total) {
+		print("Received msgNeedThisManyKits with amount " + total);
+		conveyor.msgNeedThisManyKits(total);
+		numKitsToMake = total;
+		numKitsRequested = 0;
+		kitRequested = false;
+		stateChanged();
+	}
+
+	@Override
+	public void msgKitReadyForPickup() {
+		print("Received msgKitReadyForPickup");
+		kitWaitingOnConveyor = true;
+		stateChanged();
+	}
+
+	@Override
+	public void msgNoKitsLeftOnConveyor() {
+		print("Received msgNoKitsLeftOnConveyor");
+		kitWaitingOnConveyor = false;
+		// stateChanged();
+	}
+
+	@Override
+	public void msgHereIsKit(Kit k) {
+		print("Received msgHereIsKit");
+		print("Adding kit " + k.toString());
+		kitRequested = false;
+		MyKit mk = new MyKit(k);
+		myKits.add(mk);
+		stateChanged();
+	}
+
+	@Override
+	public void msgNeedKit(int standLocation) {
+		print("Received msgNeedKit for stand location " + standLocation);
+		standPositions.put(standLocation, true);
+		stateChanged();
+	}
+
+	@Override
+	public void msgMoveKitToInspectionArea(Kit k) {
+		print("Received msgMoveKitToInspectionArea");
+		print(k.toString() + " should be moved.");
+		synchronized (myKits) {
+			// print("Acquiring in MKTIA");
+			print("Mykits size: " + myKits.size());
+			for (MyKit mk : myKits) {
+				if (mk.kit.equals(k)) {
+					mk.KS = KitStatus.MARKED_FOR_INSPECTION;
+					print("Found kit to inspect");
+					break;
+				}
+			}
+		}
+		stateChanged();
+	}
+
+	@Override
+	public void msgKitPassedInspection() {
+		print("Received msgKitPassedInspection");
+		synchronized (myKits) {
+			// print("Acquiring in KPI");
+			for (MyKit mk : myKits) {
+				if (mk.KS == KitStatus.AWAITING_INSPECTION) {
+					mk.KS = KitStatus.PASSED_INSPECTION;
+					break;
+				}
+			}
+		}
+		stateChanged();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 163:173, 178:187.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.